Ignore negative derivative when the input is total_increasing

Smankusors opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
The [documentation](https://www.home-assistant.io/integrations/derivative/) here says that we can use Utility Meter integration if we have a sensor that periodically resets to zero. But I can't get it work to output changes per second just like the original issue.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #93957
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/33152

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Smankusors committed:
if the derivative is negative, ignore it

Smankusors committed:
add option to ignore the negatives or not

Smankusors committed:
add tests for a new ignore negative derivative

Smankusors committed:
add missing description when editing

Smankusors committed:
rename to ignore_negative_derivative
to increase clarity of which negative I mean
in case in the future we want a ignore_negative_value...

elupus commented:
This should not need a configuration option. If we have a STATE_CLASS_TOTAL_INCREASING for the source sensor. We can assume to ignore negative values.

elupus commented:
What i mean is the derivative sensor should check the state class of the sensor, to know if it should ignore the negative value.

Smankusors commented:
> This should not need a configuration option. If we have a STATE_CLASS_TOTAL_INCREASING for the source sensor. We can assume to ignore negative values.
> What i mean is the derivative sensor should check the state class of the sensor, to know if it should ignore the negative value.

hmm that could work

but how about this Adguard Home sensor? It doesn't have a state class yet

![image](https://github.com/home-assistant/core/assets/10087890/d68f696e-0a78-48b3-8ee5-255262872ea3)

this sensor resets every hour, but not to zero, but to... uh... last 24 hours maybe?

will putting `total_increasing` class in this sensor make sense? 🤔

Smankusors commented:
> Well that sensor just looks weird. It resets to something strange. Just looks wrong. It should be corrected to some total sensor that reset to zero.

it seems to be just how the Adguard works I think

https://community.home-assistant.io/t/adguard-home-dns-queries-decreases-hourly/602626

elupus commented:
Well it should really be fixed in the adguard integration. Ignoring negative in integration integration is the wrong place to fix this.

frenck reviewed (commented):
Hi there @Smankusors 👋 

There is now a merge conflict on this PR (caused by another PR that was merged). Could you take a look?

Thanks! 👍 

../Frenck

Smankusors committed:
use state_class=total_increasing to ignore the negative derivative

Smankusors committed:
remove ignore negative from the config

Smankusors committed:
add test for total_increasing_reset case

Smankusors commented:
> in the meantime, for the Adguard Home sensor, is it possible to manually override it from the end user so that the sensor has a `total_increasing` class?

I figured that I could add this to the customize.yaml file, and it works 😀

```
sensor.adguard_home_dns_queries:
  state_class: total_increasing
```

---

I updated the PR now so that it now reads if the sensor is `total_increasing` instead of configuring it manually. The PR now much smaller than before 😀
wdyt? @elupus

emontnemery committed:
Merge branch 'dev' into non-negative-derivative

emontnemery commented on homeassistant/components/derivative/sensor.py:
```diff
@@ -238,6 +240,13 @@ def calc_derivative(event: Event[EventStateChangedData]) -> None:
             except AssertionError as err:
                 _LOGGER.error("Could not calculate derivative: %s", err)
 
+            if (
```
```suggestion
            # A negative derivative for a total increasing sensor means it has been
            # reset, discard the sample.
            if (
```

emontnemery commented on tests/components/derivative/test_sensor.py:
```diff
@@ -354,6 +355,31 @@ async def test_suffix(hass: HomeAssistant) -> None:
+    state = hass.states.get("sensor.power")
+    assert state is not None
+
+    assert round(float(state.state), config["sensor"]["round"]) == 2.00
```
Instead of just the magic number `2.00`, calculate it or add a comment explaining it

Smankusors replied:
at the second thought...
* I don't want to calculate it because it would duplicate the implementation in the unit tests. I want the tests to be calculation-free as much as possible so that they're independent of the algorithm.
* There's no need to comment because the test is already clear. My existing comment isn't unique to this unit test but applies to the entire component. The existing web documentation pointing to Wikipedia should suffice.

Therefore, I will remove this comment in the next commit. Wdyt?

Smankusors committed:
add comments

emontnemery commented on tests/components/derivative/test_sensor.py:
```diff
@@ -354,6 +355,34 @@ async def test_suffix(hass: HomeAssistant) -> None:
 
+async def test_total_increasing_reset(hass: HomeAssistant) -> None:
+    """Test derivative sensor state with total_increasing sensor input where it should ignore the reset value."""
+    times = [20, 30, 40]
+    values = [10, 30, 0]
```
Let's extend with another sample to test the behavior of the state update after a discarded sample

Smankusors committed:
update test_total_increasing_reset with history tests
Also remove the last comment because the test is already clear
My existing comment there isn't unique to this unit test but applies to the entire component. The existing web documentation pointing to Wikipedia should suffice.

emontnemery reviewed (approved):
LGTM, thanks @Smankusors 👍
